Revise AI section

anandmeg opened the pull request:
Update the document to improve the section on using AI for serialization. Added details on handling nested and polymorphic types.

<!-- PREVIEW-TABLE-START -->

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/standard/serialization/system-text-json/how-to.md](https://github.com/dotnet/docs/blob/2eed2262664e9c2890056336313e48d84000a6ee/docs/standard/serialization/system-text-json/how-to.md) | [How to write .NET objects as JSON (serialize)](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/how-to?branch=pr-en-us-49905) |


<!-- PREVIEW-TABLE-END -->

anandmeg committed:
Revise AI section

Update the document to improve the section on using AI for serialization. Added details on handling nested and polymorphic types.

anandmeg committed:
Fixes

anandmeg committed:
Fix anchor links

BillWagner commented on docs/standard/serialization/system-text-json/how-to.md:
```diff
@@ -108,16 +108,21 @@ Starting in .NET 9, you can also customize the indent character and size using <
 > If you use `JsonSerializerOptions` repeatedly with the same options, don't create a new `JsonSerializerOptions` instance each time you use it. Reuse the same instance for every call. For more information, see [Reuse JsonSerializerOptions instances](configure-options.md#reuse-jsonserializeroptions-instances).
 
-## Use AI to serialize to JSON
+## Use AI to serialize nested and polymorphic types
```
Those anchors have already been updated.

BillWagner reviewed (commented):
I had a couple small comments, but this is good.

I'll let @gewarren review and approve.

gewarren commented on docs/standard/serialization/system-text-json/how-to.md:
```diff
@@ -108,16 +108,21 @@ Starting in .NET 9, you can also customize the indent character and size using <
-The object contains the following fields: FirstName (string), Lastname (string), Age (int).
-Provide example output.
+Generate C# code using System.Text.Json to serialize a collection of Employee objects
+1. Define classes: Employee (with Id, Name, Department, Vehicle), Department (Name, Code), Vehicle (base), Car (Doors), Motorcycle (HasSidecar), Bicycle (IsElectric)
```
These should all end in periods.

anandmeg committed:
Address feedback

adegeo commented on docs/standard/serialization/system-text-json/how-to.md:
```diff
@@ -108,16 +108,21 @@ Starting in .NET 9, you can also customize the indent character and size using <
+3. Create a List<Employee> with sample data (one Car, one Motorcycle, one Bicycle).
+4. Serialize the list to a JSON string.
+5. Provide the complete, compiling code including all using directives, class definitions, converter implementations, and serialization logic.
+Show example output.
```
This isn't a good example, by any means. With the goal of this article being "demonstrate some basics on how to serialize something" (there are some other minor issues I have with this article in general, but I'll mention those separately 😆 )

The prompts provided should show how to accomplish the tasks in this article, and potentially tailored to the user's workspace. This suggested prompt is utterly useless for anyone who sees it, except maybe a teacher that wants to generate a comprehensive example for a student.

Even the original prompt isn't helpful.

A good prompt would be something like:

```copilot-prompt
Can you serialize my <variable-name> variable into JSON and write it out to the
file `output.json`? I want the JSON indented to look pretty
```
or
```copilot-prompt
I want to make sure my <class-name> class is serializable to JSON. Can you provide
me suggestions on what I need to improve?
```
or
```copilot-prompt
I'm serializing my <object-name> object, but when I look at the JSON output, it's
missing some properties. Can you help me fix that bug?
```

------------
I did have AI analyze this article and your suggested copilot-prompt. I asked it the following:

````prompt
I'm thinking of updating this copilot example prompt to the following:
```copilot-prompt
Generate C# code using System.Text.Json to serialize a collection of Employee objects
1. Define classes: Employee (with Id, Name, Department, Vehicle), Department (Name, Code), Vehicle (base), Car (Doors), Motorcycle (HasSidecar), Bicycle (IsElectric).
2. Use a custom JsonConverter with a $type discriminator in the output. The custom converter must implement the correct method signatures for JsonConverter<T>.
3. Create a List<Employee> with sample data (one Car, one Motorcycle, one Bicycle).
4. Serialize the list to a JSON string.
5. Provide the complete, compiling code including all using directives, class definitions, converter implementations, and serialization logic.
Show example output.
```
Can you read this article and tell me if this is actually a  good demonstration of using AI
to accomplish what this article talks about? How usable is that prompt by a user?
````

Here's the response it gave me:

#### Issues with Your Proposed Prompt
Your new prompt has several problems:

1. **Complexity Mismatch**

   The prompt requires:

   - Custom JsonConverter<T> implementation with polymorphism
   - Type discriminators
   - Multiple inheritance hierarchies
   - Complex object graphs

   This is **far beyond** what the article teaches. Custom converters are mentioned only briefly as "you can implement custom converters" with a link to a different article.

2. **Poor User Experience**

   For someone learning basic serialization:

   - The prompt is intimidating and overly technical
   - It requires understanding concepts not covered in this article (converters, polymorphism, discriminators)
   - Users following this article likely don't need this level of complexity
   - The 5-step breakdown makes it look like a programming assignment, not a helpful AI interaction

3. **Wrong Article**

   This prompt would be better suited for:

   - The custom converters article
   - An article about polymorphic serialization
   - Advanced serialization scenarios

anandmeg replied:
@adegeo, @gewarren, @BillWagner - new prompt added using one of Andy's examples. Let me know what you think.

anandmeg replied:
Thanks @adegeo. Updated.
I try all prompts and refine for results. I've made a slight change. Hope that's ok.

anandmeg commented on docs/standard/serialization/system-text-json/how-to.md:
```diff
@@ -21,7 +21,7 @@ adobe-target: true
 
 > [!TIP]
-> You can use AI assistance to [serialize to JSON](#use-ai-to-serialize-to-json).
+> You can use AI assistance to [serialize to JSON](#use-ai-to-serialize-nested-and-polymorphic-types).
```
Thanks all for your input & feedback. I'll revise the example in this article using one of your suggestions below.  
I’ll leave the advanced polymorphic example for the polymorphism article (I can draft the example and move it there, but I’ll let the SMEs review and finalize it).

anandmeg committed:
Revise AI section per feedback

anandmeg committed:
Revise

Updated prompt to simplify serialization instructions and ensure proper formatting.

adegeo reviewed (changes requested):
This is looking much better. Here are some suggestions to push this through to the finish line. Thanks!

anandmeg commented on docs/standard/serialization/system-text-json/how-to.md:
```diff
@@ -110,14 +110,14 @@ Starting in .NET 9, you can also customize the indent character and size using <
+You can use AI tools, such as GitHub Copilot, to generate code that uses `System.Text.Json` to serialize to JSON. You can customize the prompt to fit your object fields and serialization needs.
 
-The following text shows an example prompt for Copilot Chat:
+Here's an example prompt you can use in Visual Studio Code Copilot Chat to generate serialization code.
```
I tried it in Visual Studio Code.  I wouldn't want to claim it'll give the same results everywhere. User is free to copy and try the prompt with any Copilot.

anandmeg commented on docs/standard/serialization/system-text-json/how-to.md:
```diff
@@ -110,14 +110,14 @@ Starting in .NET 9, you can also customize the indent character and size using <
-Provide example output.
+I have a variable named weatherForecast of type WeatherForecast.
+Serialize the variable using System.Text.Json and write the result directly to a file stream called "output.json" with the JSON indented for pretty formatting.
+Ensure the code includes all necessary using directives and compiles without errors.
```
The difference here in using file vs file stream was that with the former it was writing to a string and then writing that string to the file.  In the latter case, it writes directly to the filestream.

anandmeg commented on docs/standard/serialization/system-text-json/how-to.md:
We used to have this disclaimer before. It was revised in the doc based on feedback we've seen to just add the current "Review Copilot's suggestions before applying them." This disclaimer is present in Copilot.

adegeo replied:
This is a requirement according to the guidelines: https://learn.microsoft.com/en-us/help/platform/copilot-in-docs-guidance#copilot-highlight-structure

adegeo replied:
The templates in the guidelines show a very simple structure for the copilot highlight, basically we don't want people focusing on the busy text outside of the prompt itself, like intros and outros to the code block. We want them at the prompt so they can be efficient at getting in and out, keeping overall copilot highlight succinct. The guideline doc has a template that can be reused, in the `templates` section, obviously 😄 but basically the template follows this pattern:

````
## Header

Simple 1 paragraph explanation

```copilot-prompt
prompt...
```
  
disclaimer
````

anandmeg commented on docs/standard/serialization/system-text-json/how-to.md:
```diff
@@ -127,4 +127,4 @@ For more information about GitHub Copilot, see GitHub's [FAQs](https://github.co
 
 - [GitHub Copilot in Visual Studio](/visualstudio/ide/visual-studio-github-copilot-install-and-states)
-- [GitHub Copilot in VS Code](https://code.visualstudio.com/docs/copilot/overview)
+- [GitHub Copilot in Visual Studio Code](https://code.visualstudio.com/docs/copilot/overview)
```
We don't want to add links directly in the section, so we decided to add these links to the Related links. wrt the other links, I'll leave it to the author to decide which ones to include.

anandmeg committed:
Apply suggestion from @adegeo

Co-authored-by: Andy (Steve) De George <67293991+adegeo@users.noreply.github.com>

anandmeg commented:
@adegeo, @gewarren - is this good to go?
cc: @mijacobs

gewarren committed:
Update docs/standard/serialization/system-text-json/how-to.md
